### PR TITLE
refactor: Cleaning up ReceiveMessage usecase

### PR DESCRIPTION
**Pull request by rjmackay**
This pull request makes the following changes:
- Ensure message is always saved
- Remove HTTP errors. They should never be in a usecase
- Remove unneeded validation
- Tidy code
- Add way more comments as I figure out whats going on

### Test script

With postman (added collection below) or with platform-client: 

- Send a message to the platform with a contact you haven't used in a targeted survey. it should be in messages but not be associated with a targeted survey. 
- Create first targeted survey (TS1) with two or more attributes, and at least 2 contacts
- After the first set of questions is sent:
    - Send a message from one of the contacts (ie using testservice) to answer the first question
    - Repeat with the same contact until its targeted_survey_state record is in survey_status "Survey finished" in the database 
    - Send another message now that it finished the survey. You should see that message in `messages` and it should not be assigned to the same post_id as the messages related to the survey
    - Send a message from the second contact to answer their first question. Do no finish the survey for this contact
- Create second targeted survey (TS2) with two or more attributes, and at least 2 contacts that are the same you used before (in TS1)
- Verify that in targeted_survey_state, the record for the contact that didn't complete the survey in TS1 reads "ACTIVE CONTACT IN SURVEY <surveyId>" where surveyId is the survey id of the first targeted survey you created
- After the first set of questions is sent, with your first contact: 
    - Send a message from one of the contacts (ie using testservice) to answer the first question
    - Verify that targeted survey has the correct attribute set affter you answer (next question's id)
    - Complete the survey with this contact. Verify that it says "Survey finished" in targeted_survey_state `survey_status` field. 
    - Send another message now that it finished the survey. You should see that message in `messages` and it should not be assigned to the same post_id as the messages related to the survey
- Send a message to the platform with a contact you haven't used in a targeted survey. it should be in messages but not be associated with a targeted survey. 

[Test targeted surveys.postman_collection.txt](https://github.com/ushahidi/platform/files/1949224/Test.targeted.surveys.postman_collection.txt)
 

Ported from #2792 
Fixes ushahidi/platform#2797

Ping @rowasc @kinstelli

### CODE REVIEW

**Comment by rjmackay** [comment]
@rowasc @kinstelli I don't know if this actually fixes anything yet. I need to write some tests before it lands... but it rebased perfectly 🎉

**Comment by rjmackay** [comment]
@rowasc @kinstelli I wrote another spec test. I found a minor bug however other than that, I'm not sure I proved anything. Spec tests are awkward for hitting all the edge cases, and mostly just showed that we need to decouple the code more.

File: src/Core/Usecase/Message/ReceiveMessage.php [hunk 1]
```diff
@@ -163,6 +163,7 @@ private function createTargetedSurveyMessages($incoming_message)
 		$messageInSurveyState = $this->repo->get($surveyStateEntity->message_id);
 
 		// If we didn't find an outgoing message
+		// I'm not convinced this is right - if someone sends multiple incoming messages before the next outgoing one. But the survey isn't finished... shouldn't we add the messages to the original post?
```
Comment (rowasc) [inline src/Core/Usecase/Message/ReceiveMessage.php]: mmm. I think this might be weird if we add it to the contact's active post for the survey since it'd be linked to the survey?

File: src/Core/Usecase/Message/ReceiveMessage.php [hunk 2]
```diff
@@ -111,118 +107,142 @@ public function setTargetedSurveyStateRepo(Entity\TargetedSurveyStateRepository
...
-		// ... attempt to load the entity
-		$messageInSurveyState = $messageInSurveyState->get($surveyStateEntity->message_id);
+		// Load the survey state for this contact
+		$surveyStateEntity = $this->targeted_survey_state_repo->getByContactId($incoming_message->contact_id);
+
+		// Attempt to load the previous message
+		$messageInSurveyState = $this->repo->get($surveyStateEntity->message_id);
+
+		// If we didn't find an outgoing message
+		// I'm not convinced this is right - if someone sends multiple incoming messages before the next outgoing one. But the survey isn't finished... shouldn't we add the messages to the original post?
```
Comment (rowasc) [inline src/Core/Usecase/Message/ReceiveMessage.php]: mmm. I think this might be weird if we add it to the contact's active post for the survey, since it'd be linked to the survey but not be a response to a question

(re-adding because I initially added the comment in a commit)

**Comment by rjmackay** [comment]
Updated to merge from develop and fix tests

File: src/Core/Usecase/Message/ReceiveMessage.php [hunk 3]
```diff
@@ -145,89 +125,129 @@ private function createIncomingMessage($incoming_message, $contact_id, $survey_s
...
 	 * @param $next_form_attribute
-	 * @return int|$incomingMessageId
-	 * @throws HTTP_Exception_400
+	 * @return int|$outgoingMessageId
 	 */
 	private function createOutgoingMessage($contact_id, $survey_state_entity, $next_form_attribute)
 	{
         // @FIXME: Message type should be configurable per deployment,survey
         $message_type = 'sms';
-        $data_provider = \DataProvider::getEnabledProviderForType($message_type);
```
Comment (rjmackay) [inline src/Core/Usecase/Message/ReceiveMessage.php]: This was breaking phpspec since I added a more thorough spec :(
I moved this handling into data providers themselves so that unassigned messages get sent via the default source

File: src/Core/Usecase/Message/ReceiveMessage.php [hunk 4]
```diff
@@ -145,89 +125,131 @@ private function createIncomingMessage($incoming_message, $contact_id, $survey_s
...
+
+		// If we have another question to send
 		if ($next_form_attribute->getId() > 0) {
-			$newMessageId = $this->createOutgoingMessage($contact_id, $surveyStateEntity, $next_form_attribute);
+			// Queue next question to be sent
+			$outgoingMessageId = $this->createOutgoingMessage(
+				$incoming_message->contact_id,
+				$surveyStateEntity,
+				$next_form_attribute,
+				$incoming_message->data_provider
```
Comment (rjmackay) [inline src/Core/Usecase/Message/ReceiveMessage.php]: We're now using the incoming data provider when sending the outgoing. ie. if we receive a message from Twilio, always reply with Twilio

**Comment by rjmackay** [comment]
@kinstelli you might want to double check my changes in the last couple of commits on this branch
I've merged anyway because this was blocking for lumen work

File: src/Core/Usecase/Message/ReceiveMessage.php (hunk 4 above)
Comment (kinstelli) [reply src/Core/Usecase/Message/ReceiveMessage.php]: Yeah, I like that better. I was thinking of doing the same thing, but was also debating about specifying SMS provider for each survey, as described in my first comment [here](https://github.com/ushahidi/platform/issues/2842), though it potentially breaks a lot of things and requires a migration to `forms`.
